src/libraries/Microsoft.Extensions.DependencyInjection/src/DependencyInjectionEventSource.cs

```diff
@@ -14,9 +16,16 @@ internal sealed class DependencyInjectionEventSource : EventSource
     {
         public static readonly DependencyInjectionEventSource Log = new DependencyInjectionEventSource();
 
-        // Event source doesn't support large payloads so we chunk formatted call site tree
+        public static class Keywords
+        {
+            public const EventKeywords ServiceProviderInitialized = (EventKeywords)0x1;
+        }
+
+        // Event source doesn't support large payloads so we chunk large payloads like formatted call site tree and descriptors
         private const int MaxChunkSize = 10 * 1024;
 
+        private readonly List<ServiceProvider> _providers = new();
```

Comment:
Do you care if this roots the `ServiceProvider`s until `DependencyInjectionEventSource.Log` gets cleaned up? That will happen in `AppContext.OnProcessEnd`. You could use a `WeakReference`, otherwise.

Reply:
Hmm maybe we should use a weak reference, that is a good idea.

Reply:
If that's the case, you'll need to build up the state if you want to ensure that you can log it at anytime while the process is up, but yes, good point.

Reply:
Done.

> you'll need to build up the state if you want to ensure that you can log it at anytime while the process is up

If someone isn't `Dispose`'ing the ServiceProvider, and it gets garbage collected, I don't believe we want to log those providers if someone attaches after the provider has been collected. We really only care about "active" ServiceProviders.